### PR TITLE
Implement auth handler and sign-in resolvers for Microsoft auth providers

### PR DESCRIPTION
**Pull request by chicoribas**
## Hey, I just made a Pull Request!

Implement auth handler and sign-in resolvers for Microsoft auth providers, as described in #6245 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Tests for new functionality and regression tests for bug fixes
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: .changeset/real-plums-vanish.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-auth-backend': minor
```
Comment (Rugvip) [inline .changeset/real-plums-vanish.md:2]: This'll be a `patch` change as it should be backwards compatible and we're on `0.x`.

File: plugins/auth-backend/api-report.md [hunk 2]
```diff
@@ -94,6 +97,20 @@ export class IdentityClient {
     }>;
     }
 
+// @public (undocumented)
+export const microsoftDefaultSignInResolver: SignInResolver<OAuthResult>;
```
Comment (Rugvip) [inline plugins/auth-backend/api-report.md:101]: Let's not export this one, as its behaviour isn't really something that we want to use in the future and we just need it for backwards compatibility

File: plugins/auth-backend/src/providers/microsoft/provider.ts [hunk 3]
```diff
@@ -131,21 +129,50 @@ export class MicrosoftAuthProvider implements OAuthHandlers {
...
+        idToken: result.params.id_token,
+        accessToken: result.accessToken,
+        scope: result.params.scope,
+        expiresInSeconds: result.params.expires_in,
       },
       profile,
-    });
+    };
+
+    if (!profile.email) {
```
Comment (Rugvip) [inline plugins/auth-backend/src/providers/microsoft/provider.ts:157]: Let's leave this check to the `authHandler`/`signInResolver`

File: plugins/auth-backend/src/providers/microsoft/provider.ts [hunk 4]
```diff
@@ -165,37 +192,102 @@ export class MicrosoftAuthProvider implements OAuthHandlers {
           resolve(photoURL);
         })
         .catch(error => {
-          console.log(
+          this.logger.warn(
```
Comment (Rugvip) [inline plugins/auth-backend/src/providers/microsoft/provider.ts:195]: 👍

File: plugins/auth-backend/src/providers/microsoft/provider.ts [hunk 5]
```diff
@@ -165,37 +192,102 @@ export class MicrosoftAuthProvider implements OAuthHandlers {
...
-    return { ...response, backstageIdentity: { id } };
+  if (!profile.email) {
+    throw new Error('Profile contained no email');
   }
-}
 
-export type MicrosoftProviderOptions = {};
+  let userId: string;
+  try {
+    const entity = await ctx.catalogIdentityClient.findUser({
```
Comment (Rugvip) [inline plugins/auth-backend/src/providers/microsoft/provider.ts:239]: The default sign-in resolver is just here for backwards compatibility and should mirror the existing logic, so we should skip the catalog query and just do the email check and split from the old `populateIdentity`

**Review by Rugvip — COMMENTED** [review]
Thank you! 👍 🎉 

Just a couple of small nits and it's good to go

**Comment by chicoribas** [comment]
Thanks fro the review @Rugvip ... 

All changes done!

**Review by Rugvip — APPROVED** [review]
Looks good! 👍

**Pull request merged**